src/lua/core/Import.cpp

```diff
@@ -48,17 +95,267 @@ static std::string get_caller(lua_State *L)
 		if (ar.source[start] == '@') {
 			start++;
 
-			static char source[1024];
-			strncpy(source, &ar.source[start], end - start + 1);
-
-			caller = std::string(source);
+			return std::string(&ar.source[start], end - start + 1);
 		}
 	}
 
 	LUA_DEBUG_END(L, 0);
 	return caller;
 }
 
+/**
+* Generate a lua module name from a given input path.
+* If the path contains a directory component containing a separator (period)
+* this function will return an empty string, as lua module names cannot contain
+* a non-separating period.
+* This function discards the trailing component of the directory path. If you
+* wish to retain it, ensure the path string ends with '/'.
+*/
+static std::string path_to_module(std::string path)
+{
+	std::string module_name;
+	// simple allocation optimizations
+	if (path.size() > 8 && path.find_first_of('/') != std::string::npos) {
+		module_name.reserve(8);
+	}
+
+	// Loop through each path fragment and append to the module name
+	// e.g. '/data/modules/path/to/file.lua' would become 'data.modules.path.to'
+	size_t start = 0;
+	size_t end = path.find_first_of('/', 1); // use start=1 to generate string::npos when the path == "/"
+
+	while (end != std::string::npos) {
+		auto fragment = path.substr(start, end - start);
+
+		if (fragment.size() > 0) {
+			// Paths containing the separator character cannot be translated into a module name
+			// this means you can't do implicit lookup using a path starting in a hidden directory, e.g.
+			// /data/.local/mylib.lua
+			if (fragment.find('.') != std::string::npos)
+				return "";
+			module_name += module_name.empty() ? fragment : '.' + fragment;
```

Comment:
In case anybody's wondering, the ternary is evaluated after the + here. I had to look it up to be sure -_-'

Two minor optimizations (not worth it here IMHO but the general pattern is worth bearing in mind) would be to

  * Not allocate a temporary string for the `'.' + fragment`, you'll need to split the operation in two appends.
  * Append the dot *after* the fragment name, unconditionally, and after the loop simply remove the extraneous '.' at the end of the string.

In the general case, the latter avoids paying the cost of testing for the initial case at each loop iteration. As I said, not really worth it here, but a nice trick nonetheless :-)

Reply:
I'm extremely used to Lua's string semantics, where strings are immutable and thus changing them by even one character triggers a copy-and-allocate operation. Under Lua's semantics it's most efficient to merge two small strings together and then into the larger string than it is to merge them individually, which would be what is tripping me up here.

Thanks for the tip though, I'll definitely use it!